Replace Phobos' parallel function with a hand-written one

ferencdg opened the pull request:
fixes #1181

Commit:
Replace Phobos' parallel function with a hand-written one

Fixes #1181

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
 
-    foreach (mod; parallel(parallel_tests))
-        runTest(mod);
+    auto available_cores = new Semaphore(totalCPUs);
```
Smart. Love it.

AndrejMitrovic reviewed (approved):
LGTM.

linked0 reviewed (approved):
LGTM

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
+    // re-initialize static variables at the start of a new task
+    void runInParallel (ModTest[] parallel_tests)
+    {
+        class WorkThread : Thread
```
~static~ Nvm

Geod24 reviewed (approved):
Nice

TrustHenry commented:
That's great! Can this make Windows parallel testing possible?

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
-        runTest(mod);
+    auto available_cores = new Semaphore(totalCPUs);
+    auto finished_tasks_num = new Semaphore(0);
+    // we cannot use phobos' parallel function, as that function will not 
```
Trailing whitespaces

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -190,11 +192,49 @@ private UnitTestResult customModuleUnitTester ()
+
+            (new WorkThread(test)).start();
+        }
+    }    
```
Trailing whitespace

Geod24 reviewed (approved):
Okay the `static` cannot be used here. Removed the trailing whitespaces (please configure your editor), and fixed capitalization of the title. LGTM, will merge when green.

ferencdg commented:
> That's great! Can this make Windows parallel testing possible?

It depends what the problem with Windows parallel testing was. We no longer call the 'parallel' function, and just depend on the Thread class

Geod24 commented:
I'm 95% certain the answer is no.

ferencdg commented:
> Okay the `static` cannot be used here. Removed the trailing whitespaces (please configure your editor), and fixed capitalization of the title. LGTM, will merge when green.

done, sorry for that
